File: app/src/main/java/com/stupidtree/hichat/ui/myprofile/MyProfileActivity.java
package com.stupidtree.hichat.ui.myprofile;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.widget.Toolbar;
import androidx.lifecycle.Observer;

import com.stupidtree.hichat.R;
import com.stupidtree.hichat.data.model.UserLocal;
import com.stupidtree.hichat.data.model.UserProfile;
import com.stupidtree.hichat.ui.base.BaseActivity;
import com.stupidtree.hichat.ui.base.DataState;
import com.stupidtree.hichat.ui.widgets.PopUpEditText;
import com.stupidtree.hichat.ui.widgets.PopUpSelectableList;
import com.stupidtree.hichat.utils.FileProviderUtils;
import com.stupidtree.hichat.utils.GalleryPicker;
import com.stupidtree.hichat.utils.ImageUtils;

import java.util.Arrays;

import butterknife.BindView;

/**
 * ”我的个人资料“ Activity
 */
public class MyProfileActivity extends BaseActivity<MyProfileViewModel> {
    /**
     * 这些是调用系统相册选择、裁剪图片要用到的状态码
     */
    public static final int RC_CHOOSE_PHOTO = 10;
    public static final int RC_TAKE_PHOTO = 11;
    public static final int RC_CROP_PHOTO = 12;

    /**
     * View绑定区
     */
    @BindView(R.id.toolbar)
    Toolbar toolbar; //toolbar

    @BindView(R.id.avatar_layout)
    ViewGroup avatarLayout;//头像那一栏

    @BindView(R.id.nickname_layout)
    ViewGroup nicknameLayout;//昵称那一栏

    @BindView(R.id.signature_layout)
    ViewGroup signatureLayout;//签名那一栏

    @BindView(R.id.nickname)
    TextView nicknameText;//昵称文本

    @BindView(R.id.signature)
    TextView signatureText;//签名文本

    @BindView(R.id.avatar)
    ImageView avatarImage;//头像图片

    @BindView(R.id.logout)
    Button logoutButton;//登出按钮

    @BindView(R.id.username)
    TextView usernameText;//用户名文本

    @BindView(R.id.gender_layout)
    ViewGroup genderLayout;//性别那一栏

    @BindView(R.id.gender)
    TextView genderText;//性别文本

    @Override
    protected Class<MyProfileViewModel> getViewModelClass() {
        return MyProfileViewModel.class;
    }


    Uri cropImgUri;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setWindowParams(true, true, false);
        setToolbarActionBack(toolbar);
        cropImgUri = Uri.parse("file:///" + Environment.getExternalStorageDirectory() + "/avatar_cropped.jpg");
    }

    @Override
    protected void initViews() {
        //点击头像那一栏，调用系统相册选择图片
        avatarLayout.setOnClickListener(view -> GalleryPicker.choosePhoto(getThis(), false));

        //当viewModel的UserProfile数据发生变更时，通知UI更新
        viewModel.getUserProfileLiveData().observe(this, userProfileDataState -> {
            if (userProfileDataState.getState() == DataState.STATE.SUCCESS) {
                setUserProfile(userProfileDataState.getData());
            } else {
                Toast.makeText(getThis(), "加载失败", Toast.LENGTH_SHORT).show();
            }

        });
        viewModel.getChangeAvatarResult().observe(this, new Observer<DataState<String>>() {
            @Override
            public void onChanged(DataState<String> stringDataState) {
                if (stringDataState.getState() == DataState.STATE.SUCCESS) {
                    Toast.makeText(getThis(), R.string.avatar_change_success, Toast.LENGTH_SHORT).show();
                    viewModel.startRefresh();
                } else {
                    Toast.makeText(getApplicationContext(), "失败", Toast.LENGTH_SHORT).show();
                }
            }
        });
        viewModel.getChangeNicknameResult().observe(this, stringDataState -> {
            if (stringDataState.getState() == DataState.STATE.SUCCESS) {
                Toast.makeText(getThis(), R.string.avatar_change_success, Toast.LENGTH_SHORT).show();
                viewModel.startRefresh();
            } else {
                Toast.makeText(getApplicationContext(), "失败", Toast.LENGTH_SHORT).show();
            }
        });

        viewModel.getChangeGenderResult().observe(this, stringDataState -> {
            if (stringDataState.getState() == DataState.STATE.SUCCESS) {
                Toast.makeText(getThis(), R.string.avatar_change_success, Toast.LENGTH_SHORT).show();
                viewModel.startRefresh();
            } else {
                Toast.makeText(getApplicationContext(), "失败", Toast.LENGTH_SHORT).show();
            }
        });

        nicknameLayout.setOnClickListener(view -> {
            DataState<UserProfile> up = viewModel.getUserProfileLiveData().getValue();
            if (up != null && up.getState() == DataState.STATE.SUCCESS) {
                new PopUpEditText()
                        .setTitle(R.string.set_nickname)
                        .setText(up.getData().getNickname())
                        .setOnConfirmListener(text -> {
                            //控制viewModel发起更改昵称请求
                            viewModel.startChangeNickname(text);
                        })
                        .show(getSupportFragmentManager(), "edit");
            }

        });

        genderLayout.setOnClickListener(view -> {
            DataState<UserProfile> up = viewModel.getUserProfileLiveData().getValue();
            if (up != null && up.getState() == DataState.STATE.SUCCESS) {
                new PopUpSelectableList<UserLocal.GENDER>()
                        .setTitle(R.string.choose_gender)
                        .setInitValue(up.getData().getGender())
                        .setListData(
                                Arrays.asList(getString(R.string.male), getString(R.string.female)),
                                Arrays.asList(UserLocal.GENDER.MALE, UserLocal.GENDER.FEMALE)
                        ).setOnConfirmListener((title, key) -> {
                    viewModel.startChangeGender(key);
                }).show(getSupportFragmentManager(), "select");
            }

        });
        logoutButton.setOnClickListener(view1 -> {
            //通知ViewModel登出
            viewModel.logout();
            finish();
        });
    }

    /**
     * 根据用户资料Model，设置UI组件
     *
     * @param profile 用户资料对象
     */
    private void setUserProfile(UserProfile profile) {
        //设置头像
        ImageUtils.loadLocalAvatarInto(getThis(), profile.getAvatar(), avatarImage);
        //设置各种文本信息
        nicknameText.setText(profile.getNickname());
        signatureText.setText(profile.getSignature());
        usernameText.setText(profile.getUsername());
        genderText.setText(profile.getGender() == UserLocal.GENDER.MALE ? R.string.male : R.string.female);
    }

    @Override
    protected void onStart() {
        super.onStart();
        //Activity启动时，就通知viewModel进行刷新UI上的用户资料
        viewModel.startRefresh();
    }

    @Override
    protected int getLayoutId() {
        return R.layout.activity_my_profile;
    }

    /**
     * 当用户通过系统相册选择图片返回时，将调用本函数
     */
    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode != RESULT_OK) {
            return;
        }
        switch (requestCode) {
            case RC_CHOOSE_PHOTO:
                //选择图片返回，要跳转到图片裁剪
                if (null == data) {
                    Toast.makeText(this, R.string.no_image_selected, Toast.LENGTH_SHORT).show();
                    return;
                }
                Uri uri = data.getData();
                if (null == uri) { //如果单个Uri为空，则可能是1:多个数据 2:没有数据
                    Toast.makeText(this, R.string.no_image_selected, Toast.LENGTH_SHORT).show();
                    return;
                }
                // 剪裁图片
                GalleryPicker.cropPhoto(getThis(), FileProviderUtils.getFilePathByUri(getThis(), uri), cropImgUri, 200);
                break;
            case RC_CROP_PHOTO:
                //裁剪图片返回，此时通知viewModel请求更改头像
                if (cropImgUri != null) {
                    String path = FileProviderUtils.getFilePathByUri(this, cropImgUri);
                    // create RequestBody instance from file
                    viewModel.startChangeAvatar(path);
                }

                break;
        }
    }

}

File: app/src/main/java/com/stupidtree/cloudliter/ui/myprofile/MyProfileActivity.kt
package com.stupidtree.cloudliter.ui.myprofile

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import com.stupidtree.cloudliter.R
import com.stupidtree.cloudliter.data.model.UserLocal
import com.stupidtree.cloudliter.data.model.UserLocal.GENDER
import com.stupidtree.cloudliter.data.model.UserProfile
import com.stupidtree.cloudliter.data.model.UserProfile.Companion.HEARING
import com.stupidtree.cloudliter.data.model.UserProfile.Companion.LIMB
import com.stupidtree.cloudliter.data.model.UserProfile.Companion.VISUAL
import com.stupidtree.cloudliter.databinding.ActivityMyProfileBinding
import com.stupidtree.style.base.BaseActivity
import com.stupidtree.component.data.DataState
import com.stupidtree.style.widgets.PopUpEditText
import com.stupidtree.style.widgets.PopUpMultipleCheckableList
import com.stupidtree.style.widgets.PopUpSelectableList
import com.stupidtree.style.picker.FileProviderUtils
import com.stupidtree.style.picker.GalleryPicker
import com.stupidtree.cloudliter.utils.ImageUtils
import com.stupidtree.cloudliter.utils.TextUtils
import com.stupidtree.style.picker.GalleryPicker.RC_CHOOSE_PHOTO
import com.stupidtree.style.picker.GalleryPicker.RC_CROP_PHOTO

/**
 * ”我的个人资料“ Activity
*/
class MyProfileActivity : BaseActivity<MyProfileViewModel, ActivityMyProfileBinding>() {

    override fun getViewModelClass(): Class<MyProfileViewModel> {
        return MyProfileViewModel::class.java
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setToolbarActionBack(binding.toolbar)

        // cropImgUri = Uri.parse("file:///" + Environment.getExternalStorageDirectory() + "/avatar_cropped.jpg")
    }

    override fun initViews() {
        //点击头像那一栏，调用系统相册选择图片
        binding.avatarLayout.setOnClickListener { GalleryPicker.choosePhoto(getThis(), false) }

        //当viewModel的UserProfile数据发生变更时，通知UI更新
        viewModel.userProfileLiveData.observe(this, { userProfileDataState: DataState<UserProfile?> ->
            if (userProfileDataState.state === DataState.STATE.SUCCESS) {
                setUserProfile(userProfileDataState.data!!)
            } else {
                Toast.makeText(getThis(), R.string.connection_failed, Toast.LENGTH_SHORT).show()
            }
        })
        viewModel.changeAvatarResult.observe(this, { stringDataState ->
            if (stringDataState.state === DataState.STATE.SUCCESS) {
                Toast.makeText(getThis(), R.string.avatar_change_success, Toast.LENGTH_SHORT).show()
                viewModel.startRefresh()
            } else {
                Toast.makeText(applicationContext, R.string.fail, Toast.LENGTH_SHORT).show()
            }
        })
        viewModel.changeNicknameResult?.observe(this, { stringDataState: DataState<String?> ->
            if (stringDataState.state === DataState.STATE.SUCCESS) {
                Toast.makeText(getThis(), R.string.avatar_change_success, Toast.LENGTH_SHORT).show()
                viewModel.startRefresh()
            } else {
                Toast.makeText(applicationContext, R.string.fail, Toast.LENGTH_SHORT).show()
            }
        })
        viewModel.changeGenderResult?.observe(this, { stringDataState: DataState<String?> ->
            if (stringDataState.state === DataState.STATE.SUCCESS) {
                Toast.makeText(getThis(), R.string.avatar_change_success, Toast.LENGTH_SHORT).show()
                viewModel.startRefresh()
            } else {
                Toast.makeText(applicationContext, R.string.fail, Toast.LENGTH_SHORT).show()
            }
        })

        viewModel.setWCAccessibilityResult?.observe(this, { stringDataState: DataState<String?> ->
            if (stringDataState.state === DataState.STATE.SUCCESS) {
                Toast.makeText(getThis(), R.string.set_success, Toast.LENGTH_SHORT).show()
                viewModel.startRefresh()
            } else {
                Toast.makeText(applicationContext, R.string.fail, Toast.LENGTH_SHORT).show()
            }
        })
        viewModel.changeAccessibilityResult?.observe(this, { stringDataState: DataState<String?> ->
            if (stringDataState.state === DataState.STATE.SUCCESS) {
                Toast.makeText(getThis(), R.string.avatar_change_success, Toast.LENGTH_SHORT).show()
                viewModel.startRefresh()
            } else {
                Toast.makeText(applicationContext, R.string.fail, Toast.LENGTH_SHORT).show()
            }
        })
        viewModel.changeTypeResult?.observe(this, { stringDataState: DataState<String?> ->
            if (stringDataState.state === DataState.STATE.SUCCESS) {
                Toast.makeText(getThis(), R.string.avatar_change_success, Toast.LENGTH_SHORT).show()
                viewModel.startRefresh()
            } else {
                Toast.makeText(applicationContext, R.string.fail, Toast.LENGTH_SHORT).show()
            }
        })
        viewModel.changeSignatureResult?.observe(this, { stringDataState: DataState<String?> ->
            if (stringDataState.state === DataState.STATE.SUCCESS) {
                Toast.makeText(getThis(), R.string.avatar_change_success, Toast.LENGTH_SHORT).show()
                viewModel.startRefresh()
            } else {
                Toast.makeText(applicationContext, R.string.fail, Toast.LENGTH_SHORT).show()
            }
        })
        binding.nicknameLayout.setOnClickListener {
            val up = viewModel.userProfileLiveData.value
            if (up != null && up.state === DataState.STATE.SUCCESS) {
                PopUpEditText()
                        .setTitle(R.string.set_nickname)
                        .setText(up.data!!.nickname)
                        .setOnConfirmListener(object : PopUpEditText.OnConfirmListener {
                            override fun OnConfirm(text: String) {
                                //控制viewModel发起更改昵称请求
                                viewModel.startChangeNickname(text)
                            }
                        })
                        .show(supportFragmentManager, "edit")
            }
        }

        //点击更改性别，弹出选择框
        binding.genderLayout.setOnClickListener {
            val up = viewModel.userProfileLiveData.value
            if (up != null && up.state === DataState.STATE.SUCCESS) {
                PopUpSelectableList<GENDER>()
                        .setTitle(R.string.choose_gender)
                        .setInitValue(up.data!!.gender)
                        .setListData(
                                listOf(getString(R.string.male), getString(R.string.female)),
                                listOf(GENDER.MALE, GENDER.FEMALE)
                        ).setOnConfirmListener(object : PopUpSelectableList.OnConfirmListener<GENDER> {

                            override fun onConfirm(title: String?, key: GENDER) {
                                viewModel.startChangeGender(key)
                            }
                        })
                        .show(supportFragmentManager, "select")
            }
        }

        // 点击更换无障碍隐私类型，弹出选择框
        binding.typePermissionLayout.setOnClickListener {
            val up = viewModel.userProfileLiveData.value
            if (up != null && up.state === DataState.STATE.SUCCESS) {
                PopUpSelectableList<UserLocal.TYPEPERMISSION>()
                        .setTitle(R.string.type_permission)
                        .setInitValue(up.data!!.typePermission)
                        .setListData(
                                listOf(getString(R.string.type_permission_public),
                                        getString(R.string.type_permission_protected),
                                        getString(R.string.type_permission_private)),
                                listOf(UserLocal.TYPEPERMISSION.PUBLIC,
                                        UserLocal.TYPEPERMISSION.PROTECTED,
                                        UserLocal.TYPEPERMISSION.PRIVATE)
                        ).setOnConfirmListener(object : PopUpSelectableList.OnConfirmListener<UserLocal.TYPEPERMISSION> {
                            override fun onConfirm(title: String?, key: UserLocal.TYPEPERMISSION) {
                                viewModel.startChangeType(up.data!!.type, up.data!!.subType, key)
                            }
                        }).show(supportFragmentManager, "select")
            }
        }

        //点击更改用户类型，弹出多选框
        binding.typeLayout.setOnClickListener {
            viewModel.userProfileLiveData.value?.data?.let {
                    PopUpMultipleCheckableList<Int>(R.string.type, 0, 0)
                            .setInitValues(it.getTypeList(it.type))
                            .setListData(
                                    listOf(getString(R.string.type_visual),
                                            getString(R.string.type_hearing),
                                            getString(R.string.type_limb)),
                                    listOf(VISUAL,
                                            HEARING,
                                            LIMB)
                            ).setOnConfirmListener(object : PopUpMultipleCheckableList.OnConfirmListener<Int> {
                                override fun onConfirm(titles: List<String?>, data: List<Int>) {
                                    var key = 0
                                    for (d in data) {
                                        key = d xor key
                                    }
                                    viewModel.startChangeType(key, it.subType, it.typePermission)
                                }
                            }).show(supportFragmentManager, "select")
            }
        }
        binding.signatureLayout.setOnClickListener {
            val up = viewModel.userProfileLiveData.value
            if (up != null && up.state === DataState.STATE.SUCCESS) {
                PopUpEditText()
                        .setTitle(R.string.choose_signature)
                        .setText(up.data!!.signature)
                        .setOnConfirmListener(object : PopUpEditText.OnConfirmListener {
                            override fun OnConfirm(text: String) {
                                //控制viewModel发起更改签名请求
                                viewModel.startChangeSignature(text)
                            }
                        })
                        .show(supportFragmentManager, "edit")
            }
        }

        binding.wordcloudLayout.setOnClickListener {
            viewModel.userProfileLiveData.value?.let {
                if (it.data != null) {
                    PopUpSelectableList<Boolean>()
                            .setTitle(R.string.choose_word_cloud_permission)
                            .setInitValue(it.data!!.wordCloudPrivate)
                            .setListData(
                                    listOf(getString(R.string.word_cloud_public),
                                            getString(R.string.word_cloud_private)),
                                    listOf(false, true))
                            .setOnConfirmListener(object : PopUpSelectableList.OnConfirmListener<Boolean> {
                                override fun onConfirm(title: String?, key: Boolean) {
                                    viewModel.startChangeWCAccessibility(key)
                                }
                            }).show(supportFragmentManager, "select")
                }
            }


        }
    }

    /**
     * 根据用户资料Model，设置UI组件
     *
     * @param profile 用户资料对象
     */
    private fun setUserProfile(profile: UserProfile) {
        //设置头像
        ImageUtils.loadAvatarInto(getThis(), profile.avatar, binding.avatar)
        //设置各种文本信息
        binding.typePermission.setText(profile.getTypePermissionName())
        binding.type.setText(profile.getTypeName())
        binding.nickname.text = profile.nickname
        binding.wordcloudAccessibility.setText(if (profile.wordCloudPrivate) R.string.word_cloud_private else R.string.word_cloud_public)
        if (!TextUtils.isEmpty(profile.signature)) {
            binding.signature.text = profile.signature
        } else {
            binding.signature.setText(R.string.place_holder_no_signature)
        }
        binding.username.text = profile.username
        binding.gender.setText(if (profile.gender == GENDER.MALE) R.string.male else R.string.female)
    }

    override fun onStart() {
        super.onStart()
        //Activity启动时，就通知viewModel进行刷新UI上的用户资料
        viewModel.startRefresh()
    }


    /**
     * 当用户通过系统相册选择图片返回时，将调用本函数
     */
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode != Activity.RESULT_OK) {
            return
        }
        when (requestCode) {

            RC_CHOOSE_PHOTO -> {
                //选择图片返回，要跳转到图片裁剪
                if (null == data) {
                    Toast.makeText(this, R.string.no_image_selected, Toast.LENGTH_SHORT).show()
                    return
                }
                val uri = data.data
                if (null == uri) { //如果单个Uri为空，则可能是1:多个数据 2:没有数据
                    Toast.makeText(this, R.string.no_image_selected, Toast.LENGTH_SHORT).show()
                    return
                }
                // 剪裁图片
                GalleryPicker.cropPhoto(getThis(), FileProviderUtils.getFilePathByUri(getThis(), uri), 200)
            }
            RC_CROP_PHOTO -> {                //裁剪图片返回，此时通知viewModel请求更改头像
                val path = GalleryPicker.getCroppedCacheDir(this)?.let { FileProviderUtils.getFilePathByUri(this, it) }
                // create RequestBody instance from file
                path?.let { viewModel.startChangeAvatar(it) }
            }
        }
    }



    override fun initViewBinding(): ActivityMyProfileBinding {
        return ActivityMyProfileBinding.inflate(layoutInflater)
    }
}